text/0000-cstr-with-ptr.md

```diff
@@ -0,0 +1,131 @@
+
+3. Modify the `CStr` and `CString` examples that use `as_ptr` to use `with_ptr` instead.
+
+    `CString::new` has the following example code:
+
+    ```rust
+    use std::ffi::CString;
+    use std::os::raw::c_char;
+
+    extern { fn puts(s: *const c_char); }
+
+    fn main() {
+        let to_print = CString::new("Hello!").unwrap();
+        unsafe {
+            puts(to_print.as_ptr());
+        }
+    }
+    ```
+
+    Under this proposal, it would change to:
+
+    ```rust
+    use std::ffi::CString;
+    use std::os::raw::c_char;
+
+    extern { fn puts(s: *const c_char); }
+
+    fn main() {
+        let to_print = CString::new("Hello!").unwrap();
+        to_print.with_ptr(|p| unsafe {
+            puts(p);
+        });
+    }
+    ```
+
+    There are nearly identical examples on `CString` and `CStr` themselves which would change in the
+    same way.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- It deprecates another stable method, which contributes to perception of API churn.
+- It adds surface area to the API of `CStr`.
+- It's still rather easy to circumvent the help and write unsound code by "leaking" the pointer out
+of the closure:
+
+    ```rust
+    let mut ptr = ptr::null();
+    {
+        let s = CString::new("foo").unwrap();
+        s.with_ptr(|p| { ptr = p; });
+    }
+    /* ptr is now dangling */
+    ```
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Do nothing. It remains very easy to write unsound code using `CString` and `CStr::as_ptr`.
+- Just make the warnings louder and the font larger in the docs for `CStr::as_ptr`. Continue to assume that people read this documentation despite the fact that the documentation examples show correct usage, but incorrect usage abounds in the ecosystem.
+- Move the `temporary_cstring_as_ptr` lint, which warns on the most common way to write said unsound
+code (calling `as_ptr` on a temporary `CString`) from Clippy to rustc.
+- Deprecate `as_ptr` and introduce a new function that does the same thing (we would have to bikeshed
+the name), but mark it `unsafe` due to the potential for unsoundness in combination with unsafe code.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Should `with_ptr` allow the closure to return a value? It could be
+
+    ```rust
+        fn with_ptr<F: FnOnce(*const c_char) -> O, O>(&self, f: F) -> O {
+            f(self.as_ptr())
+        }
+    ```
+
+    which might be convenient but would make it easier to "leak" the pointer (as easy as
+    `let ptr = s.with_ptr(|p| p);`).
+
+- Does `f(CString::new(...).unwrap().as_ptr())` actually invoke undefined behavior, if `f` doesn't store the pointer? The author's reading of the Rust reference implies that the `CString` temporary is kept alive for the entire expression, so it's fine. However, some commenters in the RFC thread have opined that the behavior of this code is unspecified at best.
```

Comment:
Sure. Temporaries stay alive at least up to the nearest arena (see https://github.com/nikomatsakis/rust-memory-model/issues/17), and function calls do not create these.